File: google/tree/populating-next-right-pointers-in-each-node.js
// Time O(N)
// Space: Last level will have N/2 nodes / space occupied by queue will be O(N) 
var connect = function (root) {
    if (!root) return root;
    const queue = [root];

    while (queue.length) {
        // len represents no of nodes at current level in tree
        const len = queue.length;
        for (let i = 0; i < len - 1; i++) {
            const curr = queue.shift();
            curr.next = queue[0];
            if (curr.left && curr.right) queue.push.apply(queue, [curr.left, curr.right]);
        }
        const last = queue.shift();
        last.next = null;
        if (last.left && last.right) queue.push.apply(queue, [last.left, last.right]);
    }
    return root;
};

/* 
Time: O(N)
Space: O(1) no additional ds is used for traversing the nodes
*/
var connect = function (root) {
    let [curr, next] = [root, root ? root.left : null];

    while (curr && next) {
        curr.left.next = curr.right;
        if (curr.next) curr.right.next = curr.next.left;

        curr = curr.next;
        if (!curr) {
            curr = next;
            next = curr.left;
        }
    }
    return root;

};

File: masteringDsAlgo/src/dsAlgo/tree/PopulatingNextRightPointersInEachNode.java
package dsAlgo.tree;

public class PopulatingNextRightPointersInEachNode {
    public Node connect(Node root) {
        Node curr = root;
        Node next = root != null ? root.left : null;

        while(curr != null && next != null) {
            curr.left.next = curr.right;
            if (curr.next != null) {
                curr.right.next = curr.next.left;
            }
            curr = curr.next;
            if (curr == null) {
                curr = next;
                next = curr.left;
            }
        }

        return root;
    }
}

class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
